dissect/target/helpers/fsutil.py

```diff
@@ -963,3 +966,96 @@ def resolve_link(
         entry = resolve_link(fs, entry, previous_links)
 
     return entry
+
+
+def open_decompress(path: TargetPath, mode: str = "rb") -> Union[BinaryIO, TextIO]:
+    """Open and decompress a file. Handles gz and bz2 files. Uncompressed files are opened as-is.
+
+    Assumes that the ``path`` exists.
+
+    Example:
+        bytes_buf = open_decompress(Path("/dir/file.gz")).read()
+
+        for line in open_decompress(Path("/dir/file.gz"), "rt"):
+            print(line)
+    """
+
+    if path.suffix == ".gz":
+        import gzip
+
+        return gzip.open(path.open(), mode)
+    elif path.suffix == ".bz2":
+        import bz2
+
+        return bz2.open(path.open(), mode)
+    else:
+        return path.open(mode)
+
+
+def reverse_readlines(fh: TextIO, chunk_size: int = io.DEFAULT_BUFFER_SIZE) -> Iterator[str]:
+    """Like iterating over a ``TextIO`` file-like object, but starting from the end of the file.
+
+    Args:
+        fh: The file-like object (opened in text mode) to iterate lines from.
+        chunk_size: The chunk size to use for iterating over lines.
+    """
+    offset = fh.seek(0, io.SEEK_END) & ((1 << 64) - 1)
+    lines = []
+
+    prev_offset = offset
+    while offset > 0:
+        if offset < chunk_size:
+            chunk_size = offset
+        offset -= chunk_size
+        fh.seek(offset)
+
+        lines = []
+        # tell() on TextIO returns a cookie which includes encode/decoder state
+        # Lower 64 bit are the file position
+        # https://peps.python.org/pep-3116/#text-i-o
+        # See TextIOWrapper._pack_cookie in _pyio.py for more detail
+        while (fh.tell() & ((1 << 64) - 1)) < prev_offset:
+            try:
+                lines.append(fh.readline())
+            except UnicodeDecodeError:
+                offset += 1
+                fh.seek(offset)
+
+        yield from reversed(lines[1:])
+        prev_offset = offset
+
+    if lines:
+        yield lines[0]
+
+
+def year_rollover_helper(
+    path: Path, re_ts: Union[str, re.Pattern], ts_format: str, tzinfo: tzinfo = timezone.utc
+) -> Iterator[tuple[datetime, str]]:
+    """Helper function for determining the correct timestamps for log files without year notation.
+
+    Supports compressed files by using :func:`open_decompress`.
+
+    Yields tuples of the parsed timestamp and the lines of the file in reverse.
+
+    Args:
+        path: A path to the log file to parse.
+        re_ts: Regex pattern for extracting the timestamp from each line.
+        ts_format: Time format specification for parsing the timestamp.
+        tzinfo: The timezone to use when parsing timestamps.
+    """
+    # Convert the mtime to the local timezone so that we get the correct year
+    current_year = from_unix(path.stat().st_mtime).astimezone(tzinfo).year
```

Comment:
```suggestion
    mtime = path.stat().st_mtime
    current_year = from_unix(mtime).astimezone(tzinfo).year
```

Makes it easier to visually parse